File: Data Structures/Tree/Union Find/UnionFind.java
package UnionFind

type unionFind struct {
	Size         int
	parent, rank []int
}

func Create(size int) *unionFind {
	uf := unionFind{
		Size:   size,
		parent: make([]int, size),
		rank:   make([]int, size),
	}
	for i := 0; i < size; i++ {
		uf.parent[i] = i
		uf.rank[i] = 1
	}
	return &uf
}

func (uf unionFind) Find(p int) int {
	for p != uf.parent[p] {
		uf.parent[p] = uf.parent[uf.parent[p]]
		p = uf.parent[p]
	}
	return p
}

func (uf unionFind) Union(p, q int) {
	rootP, rootQ := uf.Find(p), uf.Find(q)
	if rootP == rootQ {
		return
	}

	switch {
	case uf.rank[rootP] < uf.rank[rootQ]:
		uf.parent[rootP] = rootQ
	case uf.rank[rootP] > uf.rank[rootQ]:
		uf.parent[rootQ] = rootP
	default:
		uf.parent[rootP] = rootQ
		uf.rank[rootP] += 1
	}
}

func (uf unionFind) IsConnected(p, q int) bool {
	return uf.Find(p) == uf.Find(q)
}

File: src/io/innofang/kotlin/unionfind/UnionFind.kt
package io.innofang.kotlin.unionfind

/**
 * Created by Inno Fang on 2018/3/7.
 */

class UnionFind(val count: Int) {
    private val parent: IntArray = IntArray(count) { it }
    private val rank: IntArray = IntArray(count) { 1 }

    public fun find(idx: Int): Int {
        assert(idx in 0..(count - 1))
        var idx = idx
        while (idx != parent[idx]) {
            parent[idx] = parent[parent[idx]]
            idx = parent[idx]
        }
        return idx
    }

    public fun isConnected(idx1: Int, idx2: Int) = find(idx1) == find(idx2)

    public fun union(idx1: Int, idx2: Int) {
        val root1 = find(idx1)
        val root2 = find(idx2)

        if (root1 == root2) return

        when {
            rank[root1] < rank[root2] -> parent[root1] = root2
            rank[root1] > rank[root2] -> parent[root2] = root1
            else -> {
                parent[root1] = root2
                rank[root2] += 1
            }
        }
    }
}